### PR TITLE
Fix template size for managed storage / refactor cloud-install-sys-tmplt and createtmplt.sh

### PR DESCRIPTION
**Pull request by svenvogel**
## Description

* remove unused old vars and functions e.g. today, chksum...
* add function to calculate the virtual size of template for managed storage (netapp solidfire) so that the copy process is working from NFS to iSCSI correctly
* refactor cloud-install-sys-tmplt creates all settings of template.properties
* refactor createtmplt.sh shall create only the templates e.g. uncompress and install ...

## Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [X] Cleanup (Code refactoring and cleanup, that may add test cases)

## How Has This Been Tested?

tested manually in our production environment with different files like vhd, qcow2 and ova

### CODE REVIEW

**Comment by svenvogel** [comment]
@syed can you check this?

File: scripts/storage/secondary/cloud-install-sys-tmplt [hunk 1]
```diff
@@ -250,51 +232,49 @@ destcap=$(df -P $destdir | awk '{print $4}' | tail -1 )
...
+  qemuimgcmd=$(which qemu-img)
+else
+  failed 2 "Please install qemu-img command"
 fi
 
 tmpltfile=$destdir/$localfile
-tmpltsize=$(ls -l $tmpltfile| awk -F" " '{print $5}')
-
+tmpltsize=$(ls -l $tmpltfile | awk -F" " '{print $5}')
+vrtmpltsize=$($qemuimgcmd info $tmpltfile | grep -i 'virtual size' | sed -ne 's/.*(\([0-9]*\).*/\1/p' | xargs)
```
Comment (ustcweizhou) [inline scripts/storage/secondary/cloud-install-sys-tmplt]: that only applies on qcow2 images
```
if [ "$ext" == "qcow2" ]
then
    vrtmpltsize= XXXX
else
    vrtmpltsize=tmpltsize
fi
```

File: scripts/storage/secondary/cloud-install-sys-tmplt [hunk 2]
```diff
@@ -250,51 +232,49 @@ destcap=$(df -P $destdir | awk '{print $4}' | tail -1 )
...
-  echo "Failed to install system vm template $tmpltimg to $destdir: $installrslt"
-  failed 7
+if [[ $? -ne 0 ]]; then
+  failed 2 "Failed to install system vm template $tmpltimg to $destdir: $installrslt"
 fi
 
-if [ "$ext" == "ova" ]
-then
-  tar xvf $destdir/$localfile -C $destdir &> /dev/null
+if [[ $(which qemu-img) ]]; then
```
[the remark this answers is not in the document]
Comment (svenvogel) [reply scripts/storage/secondary/cloud-install-sys-tmplt]: sure we can move them up. maybe we move it into the existing  
```suggestion
if [[ $(which qemu-img) ]]; then
```

File: scripts/storage/secondary/cloud-install-sys-tmplt [hunk 3]
```diff
@@ -250,51 +232,49 @@ destcap=$(df -P $destdir | awk '{print $4}' | tail -1 )
...
+  qemuimgcmd=$(which qemu-img)
+else
+  failed 2 "Please install qemu-img command"
```
[the remark this answers is not in the document]
Comment (svenvogel) [reply scripts/storage/secondary/cloud-install-sys-tmplt]: @ustcweizhou yes possible but not consistent. 
```suggestion
  failed 2 "Please install vhd-util command"
```
the same like here and its not so easy to install... i can add this but what we do with vhd-util?
https://cbs.centos.org/koji/fileinfo?rpmID=7142&filename=/usr/bin/vhd-util

very old and not available every where...

File: scripts/storage/secondary/cloud-install-sys-tmplt (hunk 1 above)
Comment (svenvogel) [reply scripts/storage/secondary/cloud-install-sys-tmplt]: @ustcweizhou you are right i will add this...

**Comment by DaanHoogland** [comment]
/me nitpicking: "rename createtmplt.sh to createtemplt to unify the files in overall context" in the PR description (fortunately) doesn't occur, does it?

**Comment by rhtyd** [comment]
@blueorangutan package

File: scripts/storage/secondary/cloud-install-sys-tmplt [hunk 4]
```diff
@@ -42,6 +52,15 @@ dbUser="root"
 dbPassword=
 dbPort=3306
 jasypt='/usr/share/cloudstack-common/lib/jasypt-1.9.2.jar'
+tmpldescr='SystemVM Template'
+
+if [[ ! $@ =~ ^\-.+ ]]
```
[the remark this answers is not in the document]
Comment (svenvogel) [reply scripts/storage/secondary/cloud-install-sys-tmplt]: it means that Block if the block not begins with "-" is not an option and the user gets the information from the usage block

**Comment by svenvogel** [comment]
> /me nitpicking: "rename createtmplt.sh to createtemplt to unify the files in overall context" in the PR description (fortunately) doesn't occur, does it?

@DaanHoogland Hi Daan, hi thought the same. i can rename it no problem.

What do you think about the other old scripts...
image
should we remove them... not all but createvolume.sh, installiso.sh, listvmtmplt.sh and listvolume.sh ... i dont see the sense of them.

File: scripts/storage/secondary/cloud-install-sys-tmplt [hunk 5]
```diff
@@ -20,19 +19,30 @@
...
+  printf "\nUsage: %s:\n\t-m secondary storage mount point\n\t-u http url for system vm template\n\t-h hypervisor name: kvm|vmware|xenserver|hyperv|ovm3\n\t-s mgmt server secret key\n\n" $(basename $0) >&2
 }
 
 failed() {
-  echo "Installation failed"
-  exit $1
+  if [[ -z $2 ]]; then
+    :
+  else
+    echo -e $2
```
[the remark this answers is not in the document]
Comment (svenvogel) [reply scripts/storage/secondary/cloud-install-sys-tmplt]: check for a return message, if there is no one then dont send anything

**Comment by svenvogel** [comment]
@syed @ustcweizhou changes implemented... can you check

**Comment by DaanHoogland** [comment]
@svenvogel ,

> should we remove them... not all but createvolume.sh, installiso.sh, listvmtmplt.sh and listvolume.sh ... i dont see the sense of them.
Not sure if that is wise. Aren't these useful admin tools? If they are harmfully broken and never used by the system, be my guest. otherwise no harm leaving them in place.

**Comment by svenvogel** [comment]
@DaanHoogland 
> Not sure if that is wise. Aren't these useful admin tools? If they are harmfully broken and never used by the system, be my guest. otherwise no harm leaving them in place.

i added the rename and change the command in cloud-install-sys-tmplt.

should okay now...

**Comment by ustcweizhou** [comment]
@svenvogel if you rename the file, we need to update cloudstack admin guide.

**Comment by syed** [comment]
LGTM :+1:

**Comment by syed** [comment]
I don't think we use the `createtmpl` script directly as end users. We use the `cloud-install-sys-tmplt` which invokes other scripts. So as long as that is up-to-date we should be good

**Comment by svenvogel** [comment]
@ustcweizhou how @syed said 

> We use the `cloud-install-sys-tmplt` which invokes other scripts. So as long as that is up-to-date we should be good
i dont see the other command in documentation so there should nothing to do. okay?

**Comment by ustcweizhou** [comment]
@svenvogel my mistake.
For a bash script, we'd better use .sh as suffix.
createvolume.sh and listvolume.sh are still used in services/secondary-storage/server/
do not remove them.

**Comment by svenvogel** [comment]
@ustcweizhou i added the .sh to cloud-install-sys-tmplt.sh i will change the documentation. this PR should now fine.

**Comment by ustcweizhou** [comment]
@svenvogel sorryI did not say clearly.
I do not mind that cloud-install-sys-tmplt has no .sh as suffix, because it has been used for quite long time. Maybe some cloudstack users use the same command in their deployments. We'd better keep backward compatibility.
what I object is the rename or removal of the bash scripts Daan mentioned. There are two reasons (1) a bash script should  have .sh as suffix, same as .py for python scripts. (2) The scripts might be used somewhere in code, or by some users.

**Comment by svenvogel** [comment]
@ustcweizhou 

> @svenvogel sorryI did not say clearly.
> I do not mind that cloud-install-sys-tmplt has no .sh as suffix, because it has been used for quite long time. Maybe some cloudstack users use the same command in their deployments. We'd better keep backward compatibility.
> what I object is the rename or removal of the bash scripts Daan mentioned. There are two reasons (1) a bash script should have .sh as suffix, same as .py for python scripts. (2) The scripts might be used somewhere in code, or by some users.

thats not consistent. either remove all .sh from files or add them to all files. this is consistent and logical. one should not argue differently on different places.

i think with *.sh for all script or without it. backward compatibility is no argument at all. if this so no new product or change will come in the world. cloud-install-sys-tmplt is also a bash script?!?!? 

i cant agree. please choose one of them! I make the world as I like it is not a good deal i think. for me its done here.

**Comment by DennisKonrad** [comment]
@svenvogel @ustcweizhou 
I'm also in favour of cleaning this up. I see no need for multiple copies of this file

Also noone is using this script on a daily basis. So changing this to make it better will not do much harm.

**Comment by svenvogel** [comment]
@DennisKonrad @ustcweizhou @syed ready to merge now.

**Comment by rhtyd** [comment]
@DennisKonrad there are no smoketests on this PR to confirm that this PR will not add any regression(s). The merge guildelines are:
- Two LGTMs from non-authors
- Smoketests on against at least one hypervisor such a KVM, and across KVM, VMware, XenServer when changes are not hypervisor agnostic.

@blueorangutan package

**Comment by rhtyd** [comment]
Potential regression maker merged without tests.
Note - to kick tests after packaging is OK cc @PaulAngus @borisstoyanov @nvazquez and others.

**Comment by DennisKonrad** [comment]
@rhtyd How can I run the smoketests?

The second LGTM is from me ;)

**Comment by rhtyd** [comment]
@DennisKonrad you can run it manually on your local env using some nested env such as https://github.com/rhtyd/monkeybox, otherwise one can also setup Trillian https://github.com/shapeblue/Trillian

I'll kick the tests using the blueorangutan bot that kicks tests with our Trillian env.
@blueorangutan test matrix

**Comment by rhtyd** [comment]
@svenvogel @DennisKonrad sorry guys, the change fails environment deployment across KVM, VMware and XenServer so I had to revert the commit on master to unblock other PRs from testing. Kindly re-submit the change in a PR and ping me, I'll help review and test it. Thanks. cc @PaulAngus @borisstoyanov

**Comment by DaanHoogland** [comment]
> I don't think we use the `createtmpl` script directly as end users. We use the `cloud-install-sys-tmplt` which invokes other scripts. So as long as that is up-to-date we should be good

I think it is irrelevant if we have an extension or not. The issue is if it *might* be used by admions of clouds around the world we must not change it without clear migration path. consistency is not an objective, as much as backwards compatibility!

**Comment by svenvogel** [comment]
@DaanHoogland @rhtyd ok. i will create new PR and i hope you can take a look whats wrong in your test. do you have a error log whats wrong with KVM, VMware and XenServer? normal cloudstack installation should not be affected by this.

which command will used to check this in trillian or jenkins?

**Pull request merged**